### PR TITLE
feat: Add report skill feature for users

### LINKED ISSUE #1530: Add feature to report a skill
**Actual Behaviour**

Currently, there is no option to allow the user to report a skill.

**Expected Behaviour**

Add this option in the skill details screen.

**Would you like to work on the issue?**

Yes

### PR DESCRIPTION
**Pull request by codedsun**
Fixes #1530: Added feature to report skills

Changes:

Screenshots for the change: 
screenshot-1536007930850

@arundhati24 @batbrain7 Please check!

### CODE REVIEW

File: app/src/main/java/org/fossasia/susi/ai/skills/skilldetails/SkillDetailsPresenter.kt [hunk 1]
```diff
@@ -137,6 +146,20 @@ class SkillDetailsPresenter(skillDetailsFragment: SkillDetailsFragment) : ISkill
...
+
+    override fun reportSendSuccess(response: Response<ReportSkillResponse>) {
+        if (response.code() == 422) {
+            skillDetailsView?.updateSkillReportStatus(utilModel.getString(R.string.reported_already))
+        } else {
+            skillDetailsView?.updateSkillReportStatus(utilModel.getString(R.string.report_send_success))
+        }
+    }
+
+
```
Comment (arundhati24) [inline app/src/main/java/org/fossasia/susi/ai/skills/skilldetails/SkillDetailsPresenter.kt]: extra line

File: app/src/main/res/layout/fragment_skill_details.xml [hunk 2]
```diff
@@ -58,12 +58,12 @@
                 <LinearLayout
                     android:layout_width="match_parent"
                     android:layout_height="wrap_content"
+                    android:layout_marginTop="@dimen/margin_small"
                     android:background="@color/md_white_1000"
                     android:gravity="start"
                     android:orientation="horizontal"
                     android:paddingEnd="@dimen/padding_small"
                     android:paddingRight="@dimen/padding_small"
-                    android:layout_marginTop="@dimen/margin_small"
```
Comment (arundhati24) [inline app/src/main/res/layout/fragment_skill_details.xml]: unrelated change

File: app/src/main/res/layout/fragment_skill_details.xml [hunk 3]
```diff
@@ -109,12 +109,12 @@
                 android:textSize="16sp" />
 
             <TextView
+                android:id="@+id/skillDetailExample"
                 android:layout_width="wrap_content"
                 android:layout_height="wrap_content"
                 android:layout_marginTop="@dimen/margin_large"
                 android:background="@color/md_white_1000"
                 android:text="@string/examples"
-                android:id="@+id/skillDetailExample"
```
Comment (arundhati24) [inline app/src/main/res/layout/fragment_skill_details.xml]: unrelated change.

File: app/src/main/res/values/strings.xml [hunk 4]
```diff
@@ -404,4 +404,13 @@
     <string name="metric_newest">\"SUSI, what are the newest skills?\"</string>
     <string name="metrics_top_games">\"SUSI, what are your top games?\"</string>
     <string name="message_no_skills_found">No skills found.</string>
+
+    <!--Report Skill -->
+
+    <string name="report_skill">Report skill</string>
+    <string name="report_send">Send report</string>
+    <string name="report_error">Report skill unsuccessful </string>
+    <string name="report_send_success">Skill reported successfully </string>
```
Comment (arundhati24) [inline app/src/main/res/values/strings.xml]: Why an extra space at the end of the string?

File: app/src/main/res/values/strings.xml [hunk 5]
```diff
@@ -404,4 +404,13 @@
     <string name="metric_newest">\"SUSI, what are the newest skills?\"</string>
     <string name="metrics_top_games">\"SUSI, what are your top games?\"</string>
     <string name="message_no_skills_found">No skills found.</string>
+
+    <!--Report Skill -->
+
+    <string name="report_skill">Report skill</string>
+    <string name="report_send">Send report</string>
+    <string name="report_error">Report skill unsuccessful </string>
```
Comment (arundhati24) [inline app/src/main/res/values/strings.xml]: Why an extra space at the end of the string?

File: app/src/main/res/values/strings.xml [hunk 6]
```diff
@@ -404,4 +404,13 @@
...
+    <string name="report_send_success">Skill reported successfully </string>
+    <string name="reported_already">Skill already reported </string>
```
Comment (arundhati24) [inline app/src/main/res/values/strings.xml]: same

File: app/src/main/res/values/strings.xml (hunk 5 above)
Comment (arundhati24) [reply app/src/main/res/values/strings.xml]: Please change the string.
Change it to something like this "Error reporting skill. Please try again later"

File: app/src/main/res/layout/fragment_skill_details.xml (hunk 3 above)
Comment (codedsun) [reply app/src/main/res/layout/fragment_skill_details.xml]: @arundhati24 I usedm cmd+alt+l and the id moves to the first by ide so as to format properly.

File: app/src/main/res/layout/fragment_skill_details.xml (hunk 2 above)
Comment (codedsun) [reply app/src/main/res/layout/fragment_skill_details.xml]: Same

File: app/src/main/res/layout/fragment_skill_details.xml (hunk 3 above)
Comment (arundhati24) [reply app/src/main/res/layout/fragment_skill_details.xml]: Still, unrelated change. Please do not make this change here.

**Comment by codedsun** [comment]
@batbrain7 @arundhati24 Please chck!

**Comment by codedsun** [comment]
@arundhati24 @batbrain7 @iamareebjamal Please check this!

**Comment by arundhati24** [comment]
@codedsun Please change "Report skill" to "Flag as inappropriate" to make it same as Skills CMS.

File: app/src/main/java/org/fossasia/susi/ai/dataclasses/ReportSkill.kt [hunk 7]
```diff
@@ -0,0 +1,13 @@
...
+/**
+ * @author : codedsun
+ * Created on 04/09/18
+ */
+data class ReportSkill(
+        val model: String,
+        val group: String,
+        val skill: String,
+        val feedback: String,
+        val accessToken: String
```
Comment (arundhati24) [inline app/src/main/java/org/fossasia/susi/ai/dataclasses/ReportSkill.kt]: Is the key name for access token same as the variable name used here?

Comment (codedsun) [reply app/src/main/java/org/fossasia/susi/ai/dataclasses/ReportSkill.kt]: yes!

Comment (codedsun) [reply app/src/main/java/org/fossasia/susi/ai/dataclasses/ReportSkill.kt]:
```
/**
      * Report a skill
      * @param query A query  map consisting of the following key value pairs
      *              model       Model of the skill(Eg. general)
      *              group       Group of the skill(Eg. Knowledge)
      *              skill       Skill to be reported
      *              feedback    The report/msg send by the user
      *              accessToken access token of the user
      * @return the call
```

Comment (codedsun) [reply app/src/main/java/org/fossasia/susi/ai/dataclasses/ReportSkill.kt]: val map: MutableMap<String, String> = HashMap()
        map.put("access_token", query.accessToken)
        map.put("feedback", query.feedback)
        map.put("skill", query.skill)
        map.put("group", query.group)
        map.put("model", query.model)

This is the map which goes as query into retrofit instance

Comment (arundhati24) [reply app/src/main/java/org/fossasia/susi/ai/dataclasses/ReportSkill.kt]: I was not asking for this. This is all the code that you have been writing. Also, note - Here also, the key is "access_token". Please log the response that is received when you flag a skill as inappropriate. Then check that response, you'll see all the keys clearly.

Comment (codedsun) [reply app/src/main/java/org/fossasia/susi/ai/dataclasses/ReportSkill.kt]: This is the type of response that we get
package org.fossasia.susi.ai.rest.responses.susi

/**
 * Response received on reporting a skill successfully
 *
 * @author : codedsun
 * Created on 04/09/18
 */
data class ReportSkillResponse(
        val feedback: String,
        val accepted: String,
        val message: String)


What's the issue we are facing?

Comment (codedsun) [reply app/src/main/java/org/fossasia/susi/ai/dataclasses/ReportSkill.kt]: @arundhati24 I have checked that earlier, and tested this successfully where we are stuck? and what's the issue that we are facing?

Comment (arundhati24) [reply app/src/main/java/org/fossasia/susi/ai/dataclasses/ReportSkill.kt]: The issue is that you are not using the correct key name. According to me, the key for access token is "access_token" and for that you need to use serialize it.
But, to confirm, you should log the response and check it.

Comment (codedsun) [reply app/src/main/java/org/fossasia/susi/ai/dataclasses/ReportSkill.kt]: Okay @arundhati24 the params in the URL is 
```
 http://localhost:4000/cms/reportSkill.json?model=general&group=Knowledge&skill=some_inappropriate_skill&feedback=report_message&access_token=6O7cqoMbzlClxPwg1is31Tz5pjVwo3. 
```

but why do we need to change that key in data class, we are just using this as setter and getter and in SkillDetailsModel class we are already pushing the data according to the params that are accepted and their it's clearly mentioned 'access_token' as the key.

Comment (codedsun) [reply app/src/main/java/org/fossasia/susi/ai/dataclasses/ReportSkill.kt]: We have also done the same in PostFeedback class as the key their mentioned is not same as access_token that's used in the URL.

Tell me what changes should I do ?

Comment (arundhati24) [reply app/src/main/java/org/fossasia/susi/ai/dataclasses/ReportSkill.kt]: Okay @codedsun I checked it. This is the query object and not the response. I took it for the response data class. To avoid any such confusion in future, could you please change the name of the data class to ReportSkillQuery?

Comment (codedsun) [reply app/src/main/java/org/fossasia/susi/ai/dataclasses/ReportSkill.kt]: Yes sure!

**Comment by codedsun** [comment]
@arundhati24 Done the changes, check it!

**Comment by codedsun** [comment]
@arundhati24 Please check this now!

**Comment by codedsun** [comment]
@iamareebjamal Pleasse check this also!

File: app/src/main/java/org/fossasia/susi/ai/skills/skilldetails/SkillDetailsFragment.kt [hunk 8]
```diff
@@ -26,17 +28,18 @@ import org.fossasia.susi.ai.R
...
 import org.fossasia.susi.ai.dataclasses.FetchFeedbackQuery
 import org.fossasia.susi.ai.dataclasses.PostFeedback
+import org.fossasia.susi.ai.dataclasses.ReportSkillQuery
 import org.fossasia.susi.ai.helper.PrefManager
 import org.fossasia.susi.ai.helper.Utils
+import org.fossasia.susi.ai.skills.SkillsActivity
 import org.fossasia.susi.ai.rest.responses.susi.GetSkillFeedbackResponse
 import org.fossasia.susi.ai.rest.responses.susi.SkillData
 import org.fossasia.susi.ai.rest.responses.susi.Stars
-import org.fossasia.susi.ai.skills.SkillsActivity
```
Comment (arundhati24) [inline app/src/main/java/org/fossasia/susi/ai/skills/skilldetails/SkillDetailsFragment.kt]: @codedsun It's been around a month since this feature is being worked upon. I have pointed out several times in my reviews not to make such unrelated changes in PRs. Still, I find such changes here. :( 
Please, remove all these changes. And, also review your own PR before requesting a final review.

File: app/src/main/java/org/fossasia/susi/ai/data/SkillDetailsModel.kt [hunk 9]
```diff
@@ -74,6 +77,27 @@ class SkillDetailsModel : ISkillDetailsModel {
...
     }
 
+    override fun sendReport(query: ReportSkillQuery, listener: ISkillDetailsModel.OnReportSendListener) {
+        val map: MutableMap<String, String> = HashMap()
+        map.put("access_token", query.accessToken)
+        map.put("feedback", query.feedback)
+        map.put("skill", query.skill)
+        map.put("group", query.group)
+        map.put("model", query.model)
+        reportSkillResponseCall = ClientBuilder.getSusiApi().reportSkill(map)
```
Comment (arundhati24) [inline app/src/main/java/org/fossasia/susi/ai/data/SkillDetailsModel.kt]: Please create this map inside a static function in ClientBuilder.java and create the map there. For a reference, you can check the implementation of fetch feedback feature.

File: app/src/main/java/org/fossasia/susi/ai/skills/skilldetails/SkillDetailsFragment.kt (hunk 8 above)
Comment (codedsun) [reply app/src/main/java/org/fossasia/susi/ai/skills/skilldetails/SkillDetailsFragment.kt]: @arundhati24 These auto optimisation feature for android studio, does this all the time to remove the classes, which are not being used. Will be updating it again.

File: app/src/main/java/org/fossasia/susi/ai/data/SkillDetailsModel.kt (hunk 9 above)
Comment (codedsun) [reply app/src/main/java/org/fossasia/susi/ai/data/SkillDetailsModel.kt]: @arundhati24 I was asked to follow the pattern of how the feedback is send, and to pass that map values using data class. Okay, now I will move that function in ClientBuilder.java.

File: app/src/main/java/org/fossasia/susi/ai/skills/skilldetails/SkillDetailsFragment.kt (hunk 8 above)
Comment (codedsun) [reply app/src/main/java/org/fossasia/susi/ai/skills/skilldetails/SkillDetailsFragment.kt]: Still the same line is present on Line 34

**Comment by codedsun** [comment]
@arundhati24 Please check it now! Done all the changes of irrelevant changes in the imports statements.

**Comment by codedsun** [comment]
@arundhati24 Done !

**Comment by codedsun** [comment]
@arundhati24 Please check this also!

**Comment by arundhati24** [comment]
@codedsun Please update the screenshot.

**Comment by codedsun** [comment]
@arundhati24 The app is crashing as of now and the layout has no certain changes since then, it is same as above posted as far as I remember.

**Comment by codedsun** [comment]
@iamareebjamal Please check 👍

**Comment by iamareebjamal** [comment]
> The app is crashing as of now

???

**Comment by codedsun** [comment]
@iamareebjamal the app is crashing on opening the settings activity, as I was not logged in.

Update : Screenshot 
screenshot-1537373032434
